### DISCUSSION TITLE
Headless renderer

### DISCUSSION THREAD

**Discussion by Clockwork-Muse** [o3de/o3de#6986] [Q&A] [answered]
I haven't been able to find any mention of this.

Is it possible to render with o3de on a system with a GPU, but no attached monitor, such as in a renderfarm?  

On linux specifically such deployments may be to machines without a running/initialized X11 session, including docker containers (as via the various NVidia container images).  For obvious reasons, setting up a software render proxy is not viable, and many examples with dummy displays fail to work at all with more recent versions of X11 and Linux systems.

Unreal has a `RenderOffscreen` flag to enable this, requiring no special host setup (other than, in the docker case, having the NVidia runtime and using an NVidia base image).

**Comment by lmbr-pip** [comment]
Theres is a null renderer mode in O3DE but I have only ever used it for hosted servers, see --rhi=null option. 

_[Apologies, I misread your question - appreciate the clarification below]_

**Reply by Clockwork-Muse** [reply]
A null renderer would be the opposite of what I need, since a null renderer would actually render nothing (Unreal also supports this for hosted servers).

I need to render frames (and then extract them, but that's a different thing)

**Reply by olegbc1** [reply]
@jeremyong-az have you tried this and know the answer per chance? 
@Clockwork-Muse It sounds like you have an interesting use case. Would you care to share any more detail?

**Reply by Clockwork-Muse** [reply]
@olegbc1 - In my case, I'm using a game engine, deployed to a docker container on headless - that is, displayless - Ubuntu hosts, to render images for machine learning synthetic data generation.  Consider as an example scenario the Unreal plugin [AirSim](https://github.com/microsoft/AirSim).  These containers are being initialized/run from an external CI system.

Without a display connected, and especially without a login being performed, X11 is not running/authenticated, which means there is no X11 socket that can be mapped.  Fortunately, with a headless renderer, that flat-out isn't required - you just have to start the docker images with the `--gpus` flag set appropriately (ie, --gpus=all`).  Which means any gpu-enabled cloud platform can now be used as a render/streaming farm, without any special setup on their end.

This capability would also enable gpu rendering tests for this project to be performed in CI scenarios (assuming gpu-enabled runners), instead of just on developer machines, which may be beneficial.

**Reply by bhack** [reply]
> to render images for machine learning synthetic data generation

Check also the recent release of https://github.com/google-research/kubric

**Reply by Clockwork-Muse** [reply]
@bhack - Didn't know about that.
However, the problem there is it's using Cycles, the blender raytracing engine, which is comparatively slow.  The point of using a game engine is to render images at real time.

**Reply by bhack** [reply]
Yes, I know. Check the Table 1 of the related paper:
https://arxiv.org/abs/2203.03570

**Comment by wintermute-motherbrain** [comment]
@moudgils Is this something we could do?  Can we render without a display attached?  I am not sure if we can initialize the RHI backend right now without a display.  Our track view system is able to render out frames to disk but I am just not sure if there is any requirement to have a display attached.

**Reply by moudgils** [reply]
In order to do this I think we will need to do two things
1. Add support to disable window creation. We actually have a ticket for this (when null renderer is used) under Systems. We can extend this work to add a separate CL param that can disable window generation explicitly and not have it tied to null renderer. 
2. For all backends and the relevant platform we will have to check if they support swapchain creation that does not require a window handle and add support for that api. Then we can switch to that api if a window is not provided to the RHI. This work is not on the roadmap.

**Reply by olegbc1** [reply]
Thanks @moudgils! Shall we put it under SIG-triage?

**Reply by wintermute-motherbrain** [reply]
@olegbc1 looks like sig-core would need to do the Systems work and sig-graphics will need to do the backend work.  We should probably split this work.

**Reply by olegbc1** [reply]
👍 I'll create issues for this

**Reply by Clockwork-Muse** [reply]
@olegbc1 - I can't find any issues to subscribe to to track this.  Is it in some other repository?

**Comment by spham-amzn** [comment, marked as answer]
This was implemented for Linux

https://github.com/o3de/sig-simulation/issues/89
https://github.com/o3de/o3de/pull/18093

**Reply by spham-amzn** [reply]
Note however that the offline rendering is done specifically in the ROS2 gem, so as of this writing, running in console mode can handle offline rendering only through that Gem specifically. (see https://github.com/o3de/o3de-extras/blob/development/Gems/ROS2/Code/Source/Camera/ROS2CameraSystemComponent.cpp )

**Reply by Clockwork-Muse** [reply]
Need to test, but...

Just "not showing a window" isn't actually what I was going for.  One of the things I was needing to be dropped as part of this was the requirement for X11/Wayland (but not drivers).  Servers likely don't have it installed/configured (or not configured for this sort of setup in a docker situation, which is complicated).
This is possible in the general sense (Unreal and Blender both do this, via EGL autodetection, as do other 3D engines).

Presumably future work will remove the dependency on the ROS2 gem, so that it's more easily available for unrelated scenarios?

**Reply by spham-amzn** [reply]
Yeah, we did put that in for headless server. For clients, this was a first iteration but still requires X11 etc. I actually have a PoC in a private branch that adds additional support for a true headless mode for launchers, which will not load the graphics drivers, but will also disable loading any of the Atom renderers as well. (This change was more extensive as it requires a change in the build and cannot be switched on during runtime).

I haven't given much thought on how we could generalize it so that a true headless launcher can still load Atom and just the video drivers only, but my guess that it will require some changes to the Atom bootstrap component.

**Reply by spham-amzn** [reply]
Just to clarify the need for this to be a build option (and thus a separate HeadlessClientLauncher) is because on Windows, the application type must be either a `Console` or `Application` and that is determined at build time.
